### DISCUSSION TITLE
Keycloak.X Configuration

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#8501] [Archive]
Please use https://github.com/keycloak/keycloak/discussions/8654 instead

We would like to gather feedback from the community on how configuration is done in Keycloak.X. 

Are there aspects that are you find confusing or complicated? Is there some things that you'd like to configure that is missing? Any other comments and suggestions are also welcome.

**Comment by stianst** [comment]
@pedroigor FIY

**Comment by pedroigor** [comment]
To help discussions, here is the list of things we are planning to improve in the short term:

* Allow configuring the server without necessarily going through the `config` command
* Allow listing the installed providers
* Allow listing the configuration options available for each provider
* Auto-complete support in CLI
* Make easier to enable TLS
* Provide hints and usage examples for specific commands and options in CLI. Mainly for the most used ones such as database config, starting, troubleshooting, etc.
* Review our defaults to make sure they are enough for a standard production-grade deployment (clustering, TLS, proxy, etc)

We would love to get feedback about the aforementioned items as well as to get feedback about the overall experience around the CLI and if you think it is the right path for making installation, configuration, and running instances easier and less verbose than the Wildfly distribution.

See this [design document](https://github.com/keycloak/keycloak-community/blob/master/design/keycloak.x/configuration.md) for more details around Keycloak.X Configuration Model.

**Comment by pedroigor** [comment]
@stianst As you suggested, we are changing the `config` command name to `build`. See https://github.com/keycloak/keycloak/pull/8515.

**Comment by Dudeplayz** [comment]
It would be nice to have a CLI version to configure the initial admin user or setting it through a configuration file, etc.

**Reply by fabufab** [reply]
I second that. Current workaround is the following:
```bash
#!/bin/bash
set -euo pipefail

# Requires httpie to enable container readiness check
hash http

# Start containers and wait for keycloak to be responsive
docker-compose up -d
while ! http --check-status --ignore-stdin --timeout=1 GET http://localhost:8180/auth/; do   
  echo "Waiting for Keycloak"
  sleep 1
done
echo "Keycloak ready"

# Give admin user an email to enable verfication email sending
TOKEN=`http --form -v POST http://localhost:8180/auth/realms/master/protocol/openid-connect/token client_id=admin-cli username=admin password=foobar grant_type=password -b | jq -r '.access_token'`

ADMIN_ID=`http -v GET http://localhost:8180/auth/admin/realms/master/users "Authorization:Bearer $TOKEN" -b | jq -r '.[0].id'`

ADMIN_INFO='{"email": "admin@invalid"}'

echo $ADMIN_INFO | http -v PUT http://localhost:8180/auth/admin/realms/master/users/$ADMIN_ID "Authorization:Bearer $TOKEN"

http -v GET http://localhost:8180/auth/admin/realms/master/users/$ADMIN_ID "Authorization:Bearer $TOKEN"
```

**Reply by stianst** [reply]
+1 

There's two separate use-cases for that; 1) create the initial admin account, and 2) recovering access if you loose the admin password. I would like to rethink how we do those though, as I don't think it's super awesome how it works in WF distribution with the add-user-keycloak script today.

**Reply by pedroigor** [reply]
@fabufab @Dudeplayz Please, take a look at https://github.com/keycloak/keycloak/discussions/8549.

**Comment by pedroigor** [comment]
@stianst @fabufab @Dudeplayz It also makes a lot of sense for me. We have [KEYCLOAK-19412](https://issues.redhat.com/browse/KEYCLOAK-19412) to track this issue in particular.

@stianst Shall we open another discussion to start gathering requirements and ideas?

I think this one should be the start point for new ideas (such as the master admin) from which we derive discussions around a specific topic.

**Comment by pedroigor** [comment]
I would like to propose a new command for managing providers. Something like:

```
kc.sh providers [COMMAND]
```

Where some examples of command could be:

* `list`, to list all installed providers. I think very handy for troubleshooting as well as debugging the server.
* `install`, to install a provider from filesystem or even through HTTP
* `uninstall`, to uninstall a provider
* `enable`, to enable a provider
* `disable`, to disable a provider

It should also be possible to query provider configuration by:

`kc.sh providers config <id>`

I'm not 100% about the above though because we have the `show-config` command that could also be used for same purposes.

**Reply by DGuhr** [reply]
I'd say let's separate them as you suggested, giving that we have more than a hundred providers the show-config command could become very unreadable otherwise. Also, the other management commands won't make sense under `show-config`.

Perhaps it makes sense to also provide a `kc.sh show-config providers` which does the same as `kc.sh providers list` because some people might expect it when using `show-config` and as they do the same, it's only a matter to pointing both commands to the same code for printing out the list.

**Reply by stianst** [reply]
Would "providers list" also show the built-in providers, or only extensions?

**Reply by stianst** [reply]
Might also be good to support some "filtering", something like:
```
kc providers list <spi name>
kc providers config <spi name>
```
I know the last one doesn't make sense as it is, but you get the idea.

**Reply by DGuhr** [reply]
@stianst 
**Question**:
* Would "providers list" also show the built-in providers, or only extensions?:

**Answer**: I'd say all at first sight, but maybe it's a good idea to default to custom in this case? 

Somewhat like this:
`kc.sh providers list` -> only list the custom providers (or, if we can find out easily, the ones which are non-default values / configured by running e.g. `kc.sh --...`)

and then `kc.sh list all` -> list all built-in providers (could become a huge list, considering we have 114 spi implementations to date)

wdyt?

Your suggestion about filtering also makes sense for me. e.g. for hostname provider:

`kc.sh providers list hostname`would return "default" and "fixed" (and possibly some custom ones).

then running `kc.sh providers hostname default list` -> should this perhaps return the properties, like frontend-url etc., for the impl? (So, generally speaking, when there are no different provider impls, should we overload "list" with showing the properties instead of the different impls?)

---
Just a note which comes to my mind: When we config the providers also via `kc providers config` i'd suggest to use the same dash-cased subcommands syntax for the spi here as we use when we run e.g. `kc.sh start-dev --spi-action-token-handler...` to be consistent. (SPI ID is actually actionTokenHandler in this case).

**Reply by pedroigor** [reply]
I agree with all the points here. I'll send a PR to the design document with what was discussed so far and once approved start the implementation.

**Reply by pedroigor** [reply]
@stianst @DGuhr Creatde https://github.com/keycloak/keycloak/discussions/8611.

**Comment by stianst** [comment]
@pedroigor @DGuhr when we start looking at vault integration we should also support using secrets for things like DB passwords.

**Reply by DGuhr** [reply]
Generally: For sure! 
Question: Do you mean sth in the lines of https://quarkus.io/guides/vault-datasource#dynamic-database-credentials ?

**Reply by vmuzikar** [reply]
I believe this concerns SPI config in general (mainly with regards to Store.X). It should be possible to somehow use creds from vault in an SPI config. WDYT?

**Reply by pedroigor** [reply]
@stianst @DGuhr @vmuzikar I've created a thread on keycloak-dev about Vault support. See https://groups.google.com/g/keycloak-dev/c/0KqDnuKBWvA.

**Comment by up-to-you** [comment]
The main issue in current configuration i think is a lack of incremental / partial / merge support.
There are only `ignore` and `override` options.
One situation that turned me gray when we **lost all our users** after realm update, what a blessing it was not a production environment.
New CLI and modern features like Vault integration are awesome, but there is a project, which trying to solve core drawback of current engine outside of maintainers repo https://github.com/adorsys/keycloak-config-cli.
Hope we can expect it's reimplementation within official one.

**Reply by Bessonov** [reply]
[keycloak-config-cli](https://github.com/adorsys/keycloak-config-cli) is really great. I hope the way becomes a official default.
